### PR TITLE
dtoh: support generating c headers with -Hc

### PR DESCRIPTION
**Pull request by moon-chilled**

### CODE REVIEW

**Comment by moon-chilled** [comment]
Probably conflicts with #10822.  Can rebase over that if it's merged.  Also, that PR obviates part of this one; it assumes the presence of `stdint.h`, which means at least c99, which means `//` comments are fine.

**Comment by thewilsonator** [comment]
Please do.

File: src/dmd/globals.d
Comment (thewilsonator) [inline src/dmd/globals.d]: As much as my OCD would like to see this aligned, please do it in a separate PR.

File: src/dmd/globals.d
Comment (thewilsonator) [inline src/dmd/globals.d]: I think its about time to make this a struct.

File: src/dmd/globals.d
Comment (thewilsonator) [inline src/dmd/globals.d]: ditto

File: src/dmd/globals.d
Comment (thewilsonator) [inline src/dmd/globals.d]: ditto

File: src/dmd/globals.d
Comment (thewilsonator) [inline src/dmd/globals.d]: ditto

**Comment by moon-chilled** [comment]
Done

**Review by thewilsonator — COMMENTED** [review]
(It would probably be best to do these in a separate PR.)

File: src/dmd/globals.d
Comment (moon-chilled) [reply src/dmd/globals.d]: Probably out of scope for this, though?

Comment (thewilsonator) [reply src/dmd/globals.d]: For this PR yes, thats why I suggested to do it in a separate PR.

Comment (moon-chilled) [reply src/dmd/globals.d]: Oh, I see; I thought you were referring to the comment alignment when you said that

File: src/dmd/globals.d
Comment (moon-chilled) [reply src/dmd/globals.d]: Undone

File: src/dmd/mars.d
Comment (thewilsonator) [inline src/dmd/mars.d]: It seems you have a number of tabs here which the CI doesn't like

Comment (moon-chilled) [reply src/dmd/mars.d]: Oops.  I assumed all of the failures were because of output format changes.  Fixed.

File: src/dmd/dtoh.d
Comment (thewilsonator) [inline src/dmd/dtoh.d]: Why replace this? (Also this is causing the tests to fail.)

Comment (moon-chilled) [reply src/dmd/dtoh.d]: #pragma once is a non-standard compiler extension, and the only reason it's used is to make humans' lives easier.  Computer-generated output can deal.

Comment (thewilsonator) [reply src/dmd/dtoh.d]: Its supported by practically all major compilers. And you still need to change this, or change the tests to match.

**Review by wilzbach — CHANGES_REQUESTED** [review]
This should also add a runnable dshell test like it was done in the original PR to test that a C app can actually link and call D with the generated header.

**Comment by MoonlightSentinel** [comment]
`-Hc` and `HC` are hard to disambiguate at times. Maybe it should be more explicit, e.g.
- `-HC`: C headers: 
- `-HCC`: C++ headers

or

- `-HC=c`: C headers: 
- `-HC=c++`: C++ headers

**Comment by MoonlightSentinel** [comment]
Also, is it realistic for a project to generate both C and C++ headers (at once)? 
Otherwise this wouldn't require introducing three aditional switches and globals.

**Comment by wilzbach** [comment]
I like `-H=(c|c++)` very much because it might be that in the future a header file generator for other languages will be added. For example, for Objective C which is also supported by `extern` at the moment.

**Comment by moon-chilled** [comment]
I definitely think a project might want to generate multiple header files at once.  Consider a project that exposes both c and c++ interfaces, with extra functionality in the c++ interface.  The reason I made separate global flags was so that would explicitly be supported.

I think this would make a lot more sense if `-HCd` is supported and `-HCf` is not.  Then you can say `-H=c,c++,objc -Hd=somedir`, and it will generate three different files with different extensions in that dir.  (D can also be supported as a language to put after the `-H` flag, but if you don't put any languages it will still be the default.)

**Comment by moon-chilled** [comment]
Any feedback on this?  Need a finalized design before the PR can move forward.

@wilzbach @MoonlightSentinel

**Comment by MoonlightSentinel** [comment]
> I definitely think a project might want to generate multiple header files at once. Consider a project that exposes both c and c++ interfaces, with extra functionality in the c++ interface. The reason I made separate global flags was so that would explicitly be supported.

That seems reasonable.

> I think this would make a lot more sense if `-HCd` is supported and `-HCf` is not. Then you can say `-H=c,c++,objc -Hd=somedir`, and it will generate three different files with different extensions in that dir. (D can also be supported as a language to put after the `-H` flag, but if you don't put any languages it will still be the default.)

Less CLI bloat sounds great. Allhough some users might want to have separate directories for each language. But this could be solved by extending `-Hd`to accept `language=directory` pairs (similar to `-mv`). 

> One thing to consider, though, is what happens if you want to dump those headers to stdout. Should the stdout behaviour be removed and `-Hd` made mandatory? Or should `-Hd` only be mandatory if you specify multiple languages?

I don't think thats necessary. DMD should simply output those headers in a deterministic order (which is already implemented right now)

**Comment by moon-chilled** [comment]
> some users might want to have separate directories for each language. But this could be solved by extending `-Hd` to accept language=directory pairs (similar to `-mv`).

I don't think that's a worthwhile tradeoff.  If we want to provide that level of granularity, then have `-Hf` and specify a full path.  But I think if someone needs to have their headers in different dirs, they can have a script to run dmd and then move them.  The main reason to support generating multiple header files at once is to avoid the overhead of compiling multiple times.

File: src/dmd/dtoh.d
Comment (ibuclaw) [reply src/dmd/dtoh.d]: > Its supported by practically all major compilers.
And all the minor compilers (dmc) we care about for ci purposes.

Being backwards compatible with gcc-2.95 and the likes of others from that era is not a goal here.

Comment (moon-chilled) [reply src/dmd/dtoh.d]: Given a choice between using a standard language feature and a nonstandard extension (which, incidentally, has been rejected from the standard for legitimate reasons), I cannot see a reason to choose the nonstandard extension over the standard language feature.

File: src/dmd/dtoh.d
Comment (ibuclaw) [inline src/dmd/dtoh.d]: I might be having dejavu, but isn't this change already in master? Perhaps a rebase is in order so that @TurkeyMan's changes aren't mixed up with this.

Comment (TurkeyMan) [reply src/dmd/dtoh.d]: Ummm, yeah not sure why this is pulling my patch in too?
Can we make a call on my one before merging this?

File: changelog/headers.dd
Comment (TurkeyMan) [inline changelog/headers.dd]: I don't actually think this is necessary; I think we can emit compatible headers without selecting variants.
I planned to wrap the C++ symbols in `#if defined(__cplusplus)`, then they will be invisible to the C compiler, and make sure the C symbols are C compatible...

File: src/dmd/dtoh.d
Comment (moon-chilled) [reply src/dmd/dtoh.d]: Yes.  Will be done before PR is done.

File: src/dmd/dtoh.d
Comment (TurkeyMan) [reply src/dmd/dtoh.d]: `#pragma once` is a compile performance thing, that's why it's preferred. Is there any realistic potential targets where it's not supported?

**Comment by TurkeyMan** [comment]
FWIW, I intended to add C99 support too, but this is not how I'd do it. I don't think it should be a distinct mode; I think it's quite possible and not even hard to emit headers that are backward compatible with C.

File: src/dmd/dtoh.d
Comment (thewilsonator) [reply src/dmd/dtoh.d]: [No](https://en.wikipedia.org/wiki/Pragma_once#Portability)

Comment (moon-chilled) [reply src/dmd/dtoh.d]: Compilers can detect header guards and optimize accordingly, so it doesn't affect performance.

Comment (thewilsonator) [reply src/dmd/dtoh.d]: I you can name a compiler that doesn't support it, then you can change it, otherwise I'm going to block this PR in this issue.

**Review by thewilsonator — CHANGES_REQUESTED** [review]
Keep pragma once.

File: src/dmd/dtoh.d
Comment (moon-chilled) [reply src/dmd/dtoh.d]: [Lacc](https://github.com/larmel/lacc).  [scc](http://www.simple-cc.org/), and [9cc](https://github.com/rui314/9cc) do not support it.

File: changelog/headers.dd
Comment (moon-chilled) [reply changelog/headers.dd]: Probably a better idea.  How does that affect `nullptr`/`_d_null`?

File: src/dmd/dtoh.d
Comment (thewilsonator) [reply src/dmd/dtoh.d]: 9cc is no longer an active project, it successor still appears to be a personal project.

Lacc has a grand total of two contributors, with 99.7% of positive diff count going to one person.

acc has a few more but more that 76% of the commits are from one person.

Comment (moon-chilled) [reply src/dmd/dtoh.d]: Your point being?  You asked me to cite c compilers, and I did.  If you'd like an example with more contributors, look at the plan9 c compiler.

File: changelog/headers.dd
Comment (TurkeyMan) [reply changelog/headers.dd]: `nullptr` can be defined to `NULL` if `__cplusplus < magic_number`.

Comment (TurkeyMan) [reply changelog/headers.dd]: So my plan was to split `decl` into `cdecl` and `cppdecl`, and emit the symbols to one or the other appropriately, and then dump `cppdecl` below wrapped in the ifdef.

File: src/dmd/dtoh.d
Comment (TurkeyMan) [reply src/dmd/dtoh.d]: Maybe you'd be better to go and drop in a PR to support `#pragma once` into those projects :P

File: changelog/headers.dd
Comment (moon-chilled) [reply changelog/headers.dd]: > `nullptr` can be defined to `NULL` if `__cplusplus < magic_number`.

I don't like that, mainly because it leaks c++ language minutiae into c code.

Comment (TurkeyMan) [reply changelog/headers.dd]: I'm not sure what you mean? `#define` is very much a C thing.
Or if you mean just the name `nullptr`; I think it's safe to say that something close to 100% of use cases will be C++11... `extern(C++)` has some pretty strong affinities for C++11 these days. The only known case of wanting to use C++98 is bootstrapping GDC.

Comment (TurkeyMan) [reply changelog/headers.dd]: I mean, the alternative would be to use `NULL`, but I think that's a whole lot crappier personally.

Comment (moon-chilled) [reply changelog/headers.dd]: I mean, if a c file includes the (dual) header, it will get a copy of `nullptr`.

Comment (TurkeyMan) [reply changelog/headers.dd]: If that's a real concern, you can undef it before the end of the file.
It's also possible to leave that define to the including code. They may need to specify definitions of the enum macros too, so it's not unreasonable to expect them to `#define nullptr NULL` if we don't want to emit that ourselves.

File: src/dmd/dtoh.d
Comment (ibuclaw) [reply src/dmd/dtoh.d]: I only consider tinycc to be a notable "other" C compiler... and that supports `#pragma once`. :-)

**Pull request closed without merging**
